Controllery sstoolkit 14

unserializable opened the pull request:
Some PR3 tests are borken here and tests for token operations need to be added, but mainly seems to work, please verify.

General idea here is:
  * have base controller
  * have subcommands inheriting from base controller and able to define their own arguments

Token operations added are 'list' and 'login'

unserializable committed:
cherry-pick token login op from local merge branch

SSTOOLKIT-14

unserializable committed:
make config/base-dev.yaml ignored ...

for convenient branch switches

unserializable committed:
ignore pycharm's virtual env folder

unserializable committed:
Remove hard requirement of owner_id fill from ...

TokenCertificate. This needs to be reviewed and is commented
with verification TODO.

unserializable committed:
Refactor to base controller with subcommands ...

and add preliminary token listing support, since it just
so much nicer to check token status from command line
when developing.

Subcommands are now:
  token
	init

Where token command takes commands: list, login

bertvi commented:
> Some PR3 tests are borken here and tests for token operations need to be added, but mainly seems to work, please verify.
> 
> General idea here is:
> 
> * have base controller
> * have subcommands inheriting from base controller and able to define their own arguments
> 
> Token operations added are 'list' and 'login'

The idea of creating a base controller to have some common methods for all the controllers is very good.

bertvi commented:
When running the toolkit, I first get the error that module "xrdsst.resources" is not found. Here the problem is that you have created "resources" as a normal folder, not as a module. I tried creating it as a module and then it works. 

Runngin "xrdsst init" seems to work fine. Then after that running "xrdsst token list" seems also to work find as it lists tokens information, but finally, when running "xrdsst token login", I get the following error: "KeyError: software_token_id".
![pr](https://user-images.githubusercontent.com/59827771/99066665-fc769100-25b1-11eb-903a-e99f81da7950.png)

unserializable commented:
> When running the toolkit, I first get the error that module "xrdsst.resources" is not found. Here the problem is that you have information, but finally, when running "xrdsst token login", I get the following error: "KeyError: software_token_id".

Ay, 'software_token_id' is not defined in your base.yaml -- since it is also missing from the template (I only added it locally).

unserializable commented:
> When running the toolkit, I first get the error that module "xrdsst.resources" is not found. Here the problem is that you have created "resources" as a normal folder, not as a module. I tried creating it as a module and then it works.

I wonder what in our environments is causing this, it works for me both in PyCharm and on command line (virtualenv). Can you tell me how you were running ``xrdsst`` when getting that message, to reproduce?

bertvi commented:
> > When running the toolkit, I first get the error that module "xrdsst.resources" is not found. Here the problem is that you have created "resources" as a normal folder, not as a module. I tried creating it as a module and then it works.
> 
> I wonder what in our environments is causing this, it works for me both in PyCharm and on command line (virtualenv). Can you tell me how you were running `xrdsst` when getting that message, to reproduce?

I installed the xrdsst with "python setup.py install", then filled in the cofiguration file with the configuration data of my local security servers and then ran "xrdsst init". After getting the error, I tried creating a new module named "res" and copied the resource under that module and updated the files using the module and it worked.
![a](https://user-images.githubusercontent.com/59827771/99078954-e70c6180-25c7-11eb-8d46-710d7bcf68ed.png)

unserializable committed:
Add 'software_token_id' default to base.yaml

Addresses:
  https://github.com/nordic-institute/X-Road-Security-Server-toolkit/pull/4#issuecomment-726740365

unserializable commented:
> > When running the toolkit, I first get the error that module "xrdsst.resources" is not found. Here the problem is that you have information, but finally, when running "xrdsst token login", I get the following error: "KeyError: software_token_id".
> 
> Ay, 'software_token_id' is not defined in your base.yaml -- since it is also missing from the template (I only added it locally).

Addded in commit f0ae6471bc281b251f8863fad47d80cc6566b74c

unserializable commented:
> > > When running the toolkit, I first get the error that module "xrdsst.resources" is not found. Here the problem is that you have created "resources" as a normal folder, not as a module. I tried creating it as a module and then it works.
> > 
> > 
> > I wonder what in our environments is causing this, it works for me both in PyCharm and on command line (virtualenv). Can you tell me how you were running `xrdsst` when getting that message, to reproduce?
> 
> I installed the xrdsst with "python setup.py install", then filled in the cofiguration file with the configuration data of my local security servers and then ran "xrdsst init". After getting the error, I tried creating a new module named "res" and copied the resource under that module and updated the files using the module and it worked.
> ![a](https://user-images.githubusercontent.com/59827771/99078954-e70c6180-25c7-11eb-8d46-710d7bcf68ed.png)

Ok, your Python is younger than mine. The import form used is ``from xrdsst.resources.texts import texts`` and according to  https://stackoverflow.com/questions/37139786/is-init-py-not-required-for-packages-in-python-3-3 there is no need for _empty_ ``__init.py__`` since 3.3.  Cement framework in use is tested on ``Python 3.5``. Do we want to go in the direction where we try to keep Python version compatibility with lower than 3.5 version?

unserializable committed:
Retouch init controller tests

unserializable committed:
Basic token op test.

unserializable committed:
remove all empty __init.py__ files

bertvi commented on config/base.yaml:
```diff
@@ -12,6 +12,7 @@ security-server:
     owner_member_class: <MEMBER_CLASS>
     owner_member_code: <MEMBER_CODE>
     security_server_code: <SERVER_CODE>
+    software_token_id: 0 # default software token ID is zero
```
To conform with the proposed and discussed configuration style, please substitute with <SOFTWARE_TOKEN_ID>

unserializable replied:
AFAIU it is never different from zero for default software token, so rather would keep the default value, changing it as follows:

```
software_token_id: 0 # <SOFT_TOKEN_ID> -- default software token ID is zero
```
I also thought it might be better to rename it to ``default_software_token_id``, but unprefixed ``software_token_pin`` was already in place, so I decided to withhold that until we have multi-token operations in place, to see how the naming will fit then.

bertvi commented on tests/unit/test_token.py:
```diff
@@ -0,0 +1,27 @@
+import urllib3
+from xrdsst.controllers.token import TokenController
+
+class TestInit(unittest.TestCase):
```
TestInit is the test class created for testing the InitController, please use another name here, e.g. TestToken

bertvi commented on tests/unit/test_token.py:
```diff
@@ -0,0 +1,27 @@
+              'software_token_id': '0',
+              'software_token_pin': '1122'}]}
+
+    def test_token_list(self):
```
It seems you test here only the negative/failing path, please also test the happy path

unserializable replied:
Commit 6a26b61c2dbe2e03c4e001f924f4123a58ff8c62

bertvi commented on tests/unit/test_token.py:
```diff
@@ -0,0 +1,27 @@
+        token_controller.load_config = (lambda: self.ss_config)
+        self.assertRaises(urllib3.exceptions.MaxRetryError, lambda: token_controller.list())
+
+    def test_token_login(self):
```
It seems you test here only the negative/failing path, please also test the happy path

bertvi commented on xrdsst/controllers/init.py:
```diff
@@ -62,33 +43,6 @@ def initialize_server(self, config_file):
             else:
                 self.init_security_server(configuration, security_server)
 
-    @staticmethod
```
Due to the refacto of InitServerController, where you moved some base functionality to BaseController, there also now is a need for unit tests for BaseController. Fortunately the tests already exist, please move the unit tests "test_load_config", "test_load_config_exception", "test_init_logging", "test_init_logging_exception" and "test_initialize_basic_conf_values" into the new TestBase class, where you put the unit tests for testing BaseController.

unserializable replied:
Commit e16e72764934eb590cc8f588dc7ef5ecd2efd9f6

bertvi reviewed (changes requested):
Please update project version in "core/version.py" and update the project documentation related to the new token functionality in "docs/xroad_security_server_user_guide.md" and then please drop a line or two in the CHANGELOG.md file about what new has been added/changed in the latest version

unserializable committed:
Rename token test class

unserializable committed:
add doc placeholder for software_token_id

unserializable committed:
Update user documentation

unserializable committed:
bump version to 0.1.1 (alpha)

unserializable committed:
move initialization test base functinality tests to base controller unit test

unserializable committed:
Refactor: some init tests to base controller tests

unserializable committed:
Test the happy paths.
